base/traits.jl

```diff
@@ -20,3 +20,40 @@ TypeArithmetic(instance) = TypeArithmetic(typeof(instance))
 TypeArithmetic(::Type{<:AbstractFloat}) = ArithmeticRounds()
 TypeArithmetic(::Type{<:Integer}) = ArithmeticOverflows()
 TypeArithmetic(::Type{<:Any}) = ArithmeticUnknown()
+
+# trait for objects that support ranges with regular step
+"""
+    TypeRangeStep(instance)
+    TypeRangeStep(T::Type)
+
+Indicate whether an instance or a type supports constructing a range with
+a perfectly regular step or not. A regular step means that
+[`step`](@ref) will always be exaclty equal to the difference between two
+subsequent elements in a range, i.e. for a range `r::Range{T}`:
+```julia
+all(diff(r) .== step(r))
+```
+
+When a type `T` always leads to ranges with regular steps, it should
+define the following method:
+```julia
+Base.TypeRangeStep(::Type{<:Range{<:T}}) = RangeStepRegular()
```

Comment:
should we say `Base.RangeStepRegular()` since user code will more often need to qualify / import it?